AMDGPU: share LDS budget logic and add experimental LDS buffering pass

yxsamliu opened the pull request:
Add AMDGPULDSBuffering pass to buffer per-thread global memory accesses through LDS. The pass transforms load-store pairs on the same global pointer into memcpy operations through LDS (global->LDS and LDS->global). The main purpose is to alleviate global memory contention and cache thrashing when the same global pointer is used for both load and store.

This pass was inspired by finding that some rocrand performance tests show better performance when global memory is buffered through LDS instead of being loaded/stored to registers directly.

Extract a reusable LDS budget computation helper (Utils/AMDGPULDSUtils) and refactor AMDGPUPromoteAlloca to use it. This centralizes LDS usage/limit estimation including extern dynamic shared memory and local-AS args, and ties limits to occupancy tiers consistently across passes.

Gate AMDGPULDSBuffering with the same LDS budget and per-candidate accounting to avoid exceeding available LDS when multiple candidates exist. The optimization is experimental and must be enabled via the -amdgpu-enable-lds-buffering flag. It may be turned on by default later if better heuristics are created.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPU.h:
```diff
@@ -270,6 +270,21 @@ struct AMDGPUPromoteAllocaToVectorPass
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &AM);
+
+private:
+  const TargetMachine &TM;
```
Use target subclass

yxsamliu replied:
will do

arsenm commented on llvm/test/Transforms/AMDGPU/lds-buffering-basic.ll:
```diff
@@ -0,0 +1,28 @@
+  ret void
+}
+
+attributes #0 = { "amdgpu-flat-work-group-size"="1,256" "target-cpu"="gfx950" "uniform-work-group-size"="true" }
```
```suggestion
attributes #0 = { "amdgpu-flat-work-group-size"="1,256" "uniform-work-group-size"="true" }
```
Either have the attribute or the command line flag, not both

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULDSBuffering.cpp:
```diff
@@ -0,0 +1,341 @@
+    LLVM_DEBUG(dbgs() << "[LDSBuffer] Visit function: " << F.getName() << '\n');
+    const Triple &TT = TM.getTargetTriple();
+    if (!TT.isAMDGCN())
+      return false;
```
Don't add the pass in the first place

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULDSBuffering.cpp:
```diff
@@ -0,0 +1,341 @@
+
+private:
+  // Get local size Y and Z from the dispatch packet on HSA.
+  std::pair<Value *, Value *> getLocalSizeYZ(IRBuilder<> &Builder) {
```
Probably should be in utils

arsenm reviewed (changes requested):
I don't understand this transform; the load/store forwarding optimization already happens in this example and this folds to an empty function: https://godbolt.org/z/xqhc77q8c

yxsamliu commented:
> I don't understand this transform; the load/store forwarding optimization already happens in this example and this folds to an empty function: https://godbolt.org/z/xqhc77q8c

The test needs to be more complicated to reproduce the issue. A real example is in rocrand benchmark rocrand-device-lfsr113-log-normal-float-default (https://github.com/ROCm/rocm-libraries/blob/develop/projects/rocrand/benchmark/benchmark_rocrand_device_api.cpp#L56), where `states` is read, partially updated, then written back. `data` is also updated. Due to potential aliasing between `states` and `data`, compiler cannot eliminate read and write to `states`. adding LDS buffering for `states` alleviates memory contention and showed perf gains.

yxsamliu committed:
AMDGPU: add LDS buffering pass and shared LDS budget helper

Partially updating an aggregate can leave an unchanged fragment loaded
from and later stored to the same global address. Potentially aliasing
accesses can keep this pair from being removed, leaving the value live
in VGPRs.

Add an experimental pass that buffers eligible values through LDS to
shorten their VGPR live ranges. Share the LDS budget computation with
alloca promotion so both transforms respect the same occupancy limits.

shiltian reviewed (commented):
If I understand correctly, this is to enable the compiler to automatically recognize the access pattern of global pointers, put them into shared memory, and then write back. This is a pretty interesting transformation, and there are several research papers doing similar transformation as well.

However, I'm not sold by the only pattern supported at this moment. Like @arsenm mentioned earlier, the minimal pattern should have been optimized by existing optimization. If that is not the case, we should understand why instead of adding a transformation pass. If something is partially updated, how can it only have one use that is a store to the same location?

This pass can be definitely more useful if we can figure out more patterns.

yxsamliu committed:
AMDGPU: Fix LDS buffering resource accounting

Use the allocated type size and overflow-safe arithmetic when reserving per-thread LDS slots. Account for alignment padding and reject unsupported targets.\n\nShare thread index construction with promote-alloca and cover the resource and target boundaries.

yxsamliu commented:
@arsenm Re: test complexity

I added a test based on the real case. SROA splits a partly updated state object, and aliasing prevents the unchanged part from being forwarded away. The test shows the resulting load/store pair and the LDS transformation.

yxsamliu commented:
@shiltian Re: supported pattern

The one-use pair appears after SROA splits a partly updated object. An access through another pointer may alias, so existing passes cannot remove the pair. I added a test showing this flow. I kept this patch focused on the pattern seen in rocRand. I will study other useful patterns further and follow up separately.
